File: src/main/python/delta_lake.py
import time
import timeit

from delta.tables import *
from pyspark.sql import SparkSession
from pathlib import Path
import time


number_iterations = 5


def time_(function, number_iterations=number_iterations):
    start = timeit.default_timer()
    for i in range(0, number_iterations):
        function
    stop =timeit.default_timer()
    duration = (stop - start) * (10 ** 6)
    print("Time per ", number_iterations, " iterations (s) : ", duration)
    print("Time per  single iteration (s) : ", duration / number_iterations)


class Delta_lake:
    def __init__(self, display):
        self.spark = SparkSession.builder.appName("Delta_crud") \
            .config("spark.jars.packages", "io.delta:delta-core_2.12:0.7.0") \
            .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension") \
            .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog") \
            .getOrCreate()

        self.sc = self.spark.sparkContext
        self.sc.setLogLevel("ERROR")
        self.pwd = str(Path(__file__).parents[3])
        self.path_delta_table = self.pwd + "\\data\\delta_database"
        self.path_data = self.pwd + "\\data\\brut_data\\data.parquet"
        self.delta_table = None
        self.df = self.spark.read.load(self.path_data)
        self.display_ = display

    def read(self):
        self.delta_table = DeltaTable.forPath(self.spark, self.path_delta_table)

    def write(self):
        self.df.write.format("delta").mode("overwrite") \
            .save(self.path_delta_table)
        self.read()

    def insert(self):
        df_insert = self.df.limit(1)
        self.delta_table.alias("people").merge(
            df_insert.alias("updates"),
            "people.id = updates.id") \
            .whenMatchedUpdate(set={"id": "updates.id"}) \
            .whenNotMatchedInsert(values=
                                  {"registration_dttm": "updates.registration_dttm",
                                   "id": "updates.id",
                                   "first_name": "updates.first_name",
                                   "last_name": "updates.last_name",
                                   "email": "updates.email",
                                   "gender": "updates.gender",
                                   "ip_address": "updates.ip_address",
                                   "cc": "updates.cc",
                                   "country": "updates.country",
                                   "birthdate": "updates.birthdate",
                                   "salary": "updates.salary",
                                   "title": "updates.title",
                                   "comments": "updates.comments"
                                   }).execute()
        if self.display_:
            print("+++++++++++++++++++++++++++++++++++++++++++ Insert +++++++++++++++++++++++++++++++++++++++++++")
            self.display()

    def update(self):
        self.delta_table.update(set={"id": "id + 1000"})
        if self.display_:
            print("+++++++++++++++++++++++++++++++++++++++++++ Update +++++++++++++++++++++++++++++++++++++++++++")
            self.display()

    def delete(self):
        self.delta_table.delete(condition="id == 1")
        if self.display_:
            print("+++++++++++++++++++++++++++++++++++++++++++ Delete +++++++++++++++++++++++++++++++++++++++++++")
            self.display()

    def roll_back(self):
        self.df = self.spark.read.format("delta").option("versionAsOf", 0).load(self.path_delta_table)
        if self.display_:
            print("+++++++++++++++++++++++++++++++++++++++++++ Roll_back +++++++++++++++++++++++++++++++++++++++++++")
            self.display()

    def display(self):
        self.delta_table.toDF().show()
        print("***** Size Table:", self.delta_table.toDF().count())

    def __await__(self):
        time.sleep(300)


def calculate_time():
    delta_lake = Delta_lake(False)
    delta_lake.read()
    print("+++++++++++++++++++++++++++++++ Writing +++++++++++++++++++++++++++++++")
    time_(delta_lake.write())
    print("+++++++++++++++++++++++++++++++ Updating +++++++++++++++++++++++++++++++")
    time_(delta_lake.update())
    print("+++++++++++++++++++++++++++++++ Inserting +++++++++++++++++++++++++++++++")
    time_(delta_lake.insert())
    print("+++++++++++++++++++++++++++++++ Deleting  +++++++++++++++++++++++++++++++")
    time_(delta_lake.delete())


def test():
    delta_lake = Delta_lake(True)
    delta_lake.write()
    delta_lake.update()
    delta_lake.insert()
    delta_lake.delete()


if __name__ == '__main__':
    calculate_time()

File: src/main/scala/Delta_lake.scala

import io.delta.tables.DeltaTable
import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.{SparkConf, SparkContext}


object Delta_lake {
  var path_data: String = _
  var path_delta_table: String = _
  var conf: SparkConf = _
  var sc: SparkContext = _
  var spark: SparkSession = _
  var people_df: DataFrame = _
  var pwd: String = _
  var number_iterations = 5
  var delta_table: DeltaTable = _


  def time[T](f: => T, number_iterations: Int = 1): Unit = {
    val time_init = System.nanoTime
    for (i <- 1 to number_iterations) {
      var function_call = f
      function_call
    }
    val duration = (System.nanoTime - time_init) / 1e9d
    println("Time per " + number_iterations + " iterations (s) : " + duration)
    println("Time per  single iteration (s) : " + duration / number_iterations)
  }

  def init_spark_session(): Unit = {
    pwd = System.getProperty("user.dir")
    path_data = pwd + "\\data\\brut_data\\data.parquet"
    path_delta_table = pwd + "\\data\\delta_database"
    conf = new SparkConf().setMaster("local").setAppName("Delta_operation")
    conf.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
    sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")
    spark = SparkSession.builder.config(sc.getConf).getOrCreate()
    people_df = spark.read.parquet(path_data)

  }

  def read(): Unit = {
    delta_table = DeltaTable.forPath(spark, path_delta_table)
  }

  def display(): Unit = {
    people_df.show()
    println("Size df : " + people_df.count())
  }


  def insert(): Unit = {
    var df_insert = people_df.limit(1)
    delta_table.alias("people").merge(df_insert.alias("updates"), "people.id = updates.id")
      .whenNotMatched
      .insertExpr(
        Map("registration_dttm" -> "updates.registration_dttm",
          "id" -> "updates.id",
          "first_name" -> "updates.first_name",
          "last_name" -> "updates.last_name",
          "email" -> "updates.email",
          "gender" -> "updates.gender",
          "ip_address" -> "updates.ip_address",
          "cc" -> "updates.cc",
          "country" -> "updates.country",
          "birthdate" -> "updates.birthdate",
          "salary" -> "updates.salary",
          "title" -> "updates.title",
          "comments" -> "updates.comments"
        ))
      .execute()


  }

  def write(): Unit = {
    people_df.write.format("delta").mode("overwrite")
      .save(path_delta_table)
  }

  def update(): Unit = {

    delta_table.updateExpr(
      Map("id" -> "id + 1000 "))
  }

  def delete(): Unit = {
    delta_table.delete(condition = "id == 1")
  }

  def calculate_time(): Unit = {
    read()
    println("+++++++++++++++++++++++++++++++ Writing +++++++++++++++++++++++++++++++")
    time(write(), number_iterations)
    println("+++++++++++++++++++++++++++++++ Updating +++++++++++++++++++++++++++++++")
    time(update(), number_iterations)
    println("+++++++++++++++++++++++++++++++ Inserting +++++++++++++++++++++++++++++++")
    time(insert(), number_iterations)
    println("+++++++++++++++++++++++++++++++ Deleting  +++++++++++++++++++++++++++++++")
    time(delete(), number_iterations)


  }

  def main(args: Array[String]): Unit = {
    init_spark_session()
    calculate_time()

  }
}
